File: nodejs/src/gateways/fib.ts
/**
 * FIB (First Iraqi Bank) Gateway Client for Node.js
 * Handles OAuth2 authentication and payment operations
 */

import axios, { AxiosInstance } from 'axios';
import { PaymentRequest, PaymentResponse, PaymentStatus, PaymentProvider } from '../index';

export interface FIBConfig {
  clientId: string;
  clientSecret: string;
  baseUrl: string;
}

export interface FIBTokenResponse {
  access_token: string;
  token_type: string;
  expires_in: number;
  refresh_token?: string;
}

export interface FIBTokenInfo {
  accessToken: string;
  expiresAt: number;
  refreshToken?: string;
}

export interface FIBPaymentRequest {
  amount: number;
  order_id: string;
  callback_url: string;
  webhook_url: string;
  metadata?: Record<string, string>;
}

export interface FIBPaymentResponse {
  payment_id: string;
  status: 'pending' | 'completed' | 'failed' | 'expired';
  redirect_url: string;
  deep_link?: string;
}

/**
 * FIB Gateway Client with OAuth2 token management
 */
export class FIBClient {
  private client: AxiosInstance;
  private config: FIBConfig;
  private tokenInfo: FIBTokenInfo | null = null;

  constructor(config: FIBConfig) {
    this.config = config;
    this.client = axios.create({
      baseURL: config.baseUrl,
      headers: {
        'Content-Type': 'application/json',
        'User-Agent': 'MesopotamiaSDK/0.1.0',
      },
      timeout: 30000,
    });
  }

  /**
   * Validate amount for FIB (1K - 100M IQD)
   */
  private validateAmount(amount: number): void {
    const min = 1000;
    const max = 100000000;
    if (amount < min || amount > max) {
      throw new Error(`Amount must be between ${min} and ${max} IQD`);
    }
  }

  /**
   * Get OAuth2 access token with caching
   */
  private async getAccessToken(): Promise<string> {
    // Check if we have a valid cached token
    if (this.tokenInfo && Date.now() < this.tokenInfo.expiresAt - 300000) {
      // Token is valid and not expiring within 5 minutes
      return this.tokenInfo.accessToken;
    }

    // Fetch new token
    return await this.fetchToken();
  }

  /**
   * Fetch new OAuth2 token from FIB
   */
  private async fetchToken(): Promise<string> {
    try {
      const params = new URLSearchParams();
      params.append('grant_type', 'client_credentials');
      params.append('client_id', this.config.clientId);
      params.append('client_secret', this.config.clientSecret);

      const response = await this.client.post<FIBTokenResponse>(
        '/oauth/token',
        params.toString(),
        {
          headers: {
            'Content-Type': 'application/x-www-form-urlencoded',
          },
        }
      );

      const data = response.data;

      // Cache token with expiry
      this.tokenInfo = {
        accessToken: data.access_token,
        expiresAt: Date.now() + data.expires_in * 1000,
        refreshToken: data.refresh_token,
      };

      return data.access_token;
    } catch (error: any) {
      throw new Error(`FIB OAuth token fetch failed: ${error.message}`);
    }
  }

  /**
   * Create a payment transaction
   */
  async createPayment(request: PaymentRequest): Promise<PaymentResponse> {
    this.validateAmount(request.amount);

    const accessToken = await this.getAccessToken();

    const body: FIBPaymentRequest = {
      amount: request.amount,
      order_id: request.orderId,
      callback_url: request.callbackUrl,
      webhook_url: request.webhookUrl,
      metadata: request.metadata,
    };

    try {
      const response = await this.client.post<FIBPaymentResponse>(
        '/payments',
        body,
        {
          headers: {
            'Authorization': `Bearer ${accessToken}`,
          },
        }
      );

      const data = response.data;

      return {
        transactionId: data.payment_id,
        redirectUrl: data.redirect_url,
        deepLink: data.deep_link,
        status: this.mapStatus(data.status),
        provider: PaymentProvider.FIB,
      };
    } catch (error: any) {
      // If unauthorized, clear token and retry once
      if (error.response?.status === 401 && this.tokenInfo) {
        this.tokenInfo = null;
        const newToken = await this.getAccessToken();
        const retryResponse = await this.client.post<FIBPaymentResponse>(
          '/payments',
          body,
          {
            headers: {
              'Authorization': `Bearer ${newToken}`,
            },
          }
        );

        const data = retryResponse.data;
        return {
          transactionId: data.payment_id,
          redirectUrl: data.redirect_url,
          deepLink: data.deep_link,
          status: this.mapStatus(data.status),
          provider: PaymentProvider.FIB,
        };
      }

      throw new Error(`FIB payment creation failed: ${error.message}`);
    }
  }

  /**
   * Get payment status
   */
  async getPaymentStatus(transactionId: string): Promise<PaymentStatus> {
    const accessToken = await this.getAccessToken();

    try {
      const response = await this.client.get<{status: string}>(
        `/payments/${transactionId}/status`,
        {
          headers: {
            'Authorization': `Bearer ${accessToken}`,
          },
        }
      );
      return this.mapStatus(response.data.status);
    } catch (error: any) {
      throw new Error(`FIB status check failed: ${error.message}`);
    }
  }

  /**
   * Verify webhook signature (HMAC-SHA256)
   */
  verifyWebhook(signature: string, payload: string, secret: string): boolean {
    const crypto = require('crypto');
    const sigBytes = signature.startsWith('sha256=')
      ? signature.slice(7)
      : signature;

    const expected = crypto
      .createHmac('sha256', secret)
      .update(payload)
      .digest('hex');

    return crypto.timingSafeEqual(
      Buffer.from(sigBytes),
      Buffer.from(expected)
    );
  }

  /**
   * Clear cached token (use for testing or manual refresh)
   */
  clearToken(): void {
    this.tokenInfo = null;
  }

  private mapStatus(status: string): PaymentStatus {
    switch (status) {
      case 'completed':
        return PaymentStatus.COMPLETED;
      case 'failed':
        return PaymentStatus.FAILED;
      case 'expired':
        return PaymentStatus.EXPIRED;
      default:
        return PaymentStatus.PENDING;
    }
  }
}

export default FIBClient;

File: core/src/fib.rs
// FIB (First Iraqi Bank) gateway implementation

use crate::client::HttpClient;
use crate::crypto::generate_hmac_sha256;
use crate::error::{MesopotamiaError, Result};
use crate::types::{Environment, PaymentProvider, PaymentRequest, PaymentResponse, PaymentStatus};
use reqwest::header::{AUTHORIZATION, CONTENT_TYPE};
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::sync::{Arc, Mutex};
use std::time::{SystemTime, UNIX_EPOCH};

/// FIB gateway client with OAuth2 token caching
pub struct FIBClient {
    base_url: String,
    client_id: String,
    client_secret: String,
    client: HttpClient,
    token_cache: Arc<Mutex<Option<FIBTokenInfo>>>,
}

impl FIBClient {
    /// Create a new FIB client
    pub fn new(base_url: String, client_id: String, client_secret: String, timeout_ms: u64) -> Self {
        Self {
            base_url,
            client_id,
            client_secret,
            client: HttpClient::new(timeout_ms),
            token_cache: Arc::new(Mutex::new(None)),
        }
    }

    /// Create a payment transaction
    pub async fn create_payment(&self, request: &PaymentRequest) -> Result<PaymentResponse> {
        // Validate amount range
        let (min, max) = PaymentProvider::FIB.amount_range();
        if request.amount < min || request.amount > max {
            return Err(MesopotamiaError::InvalidAmount {
                min,
                max,
                provided: request.amount,
            });
        }

        // Validate order ID
        if request.order_id.len() > 50 {
            return Err(MesopotamiaError::InvalidOrderId {
                reason: "Order ID must be max 50 characters".to_string(),
            });
        }

        // Get or refresh access token
        let access_token = self.get_access_token().await?;

        // Build request body
        let body = FIBCreatePaymentRequest {
            amount: request.amount,
            order_id: request.order_id.clone(),
            callback_url: request.callback_url.clone(),
            webhook_url: request.webhook_url.clone(),
            description: request.description.clone().unwrap_or_default(),
            metadata: request.metadata.clone().unwrap_or_default(),
        };

        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .post(format!("{}/payment/create", self.base_url))
                .header(AUTHORIZATION, format!("Bearer {}", access_token))
                .json(&body)
                .send()
                .await?;

            Ok(resp)
        }).await?;

        // Handle response
        let status = response.status();
        let response_text = response.text().await?;

        if status.is_success() {
            let create_response: FIBCreatePaymentResponse = serde_json::from_str(&response_text)
                .map_err(|e| MesopotamiaError::SerializationError(e))?;

            Ok(PaymentResponse {
                transaction_id: create_response.payment_id,
                redirect_url: create_response.web_url,
                deep_link: Some(create_response.deep_link),
                status: PaymentStatus::Pending,
                provider: PaymentProvider::FIB,
            })
        } else {
            self.handle_error(status, &response_text)
        }
    }

    /// Get payment status
    pub async fn get_payment_status(&self, payment_id: &str) -> Result<PaymentStatus> {
        let access_token = self.get_access_token().await?;

        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .get(format!("{}/payment/status/{}", self.base_url, payment_id))
                .header(AUTHORIZATION, format!("Bearer {}", access_token))
                .send()
                .await?;

            Ok(resp)
        }).await?;

        let status_response: FIBPaymentStatusResponse = response.json().await?;
        Ok(parse_status(&status_response.status))
    }

    /// Get or refresh OAuth2 access token
    async fn get_access_token(&self) -> Result<String> {
        // Check if we have a valid cached token
        {
            let cache = self.token_cache.lock().unwrap();
            if let Some(ref token_info) = *cache {
                if token_info.is_valid() {
                    return Ok(token_info.access_token.clone());
                }
            }
        }

        // Need to fetch a new token
        let token_info = self.fetch_token().await?;

        // Cache the token
        {
            let mut cache = self.token_cache.lock().unwrap();
            *cache = Some(token_info.clone());
        }

        Ok(token_info.access_token)
    }

    /// Fetch new OAuth2 token from FIB
    async fn fetch_token(&self) -> Result<FIBTokenInfo> {
        let response = self.client.execute_with_retry(|| async {
            let client = self.client.inner();
            let resp = client
                .post(format!("{}/oauth/token", self.base_url))
                .header(CONTENT_TYPE, "application/x-www-form-urlencoded")
                .form(&[
                    ("grant_type", "client_credentials"),
                    ("client_id", &self.client_id),
                    ("client_secret", &self.client_secret),
                ])
                .send()
                .await?;

            Ok(resp)
        }).await?;

        let status = response.status();
        let response_text = response.text().await?;

        if !status.is_success() {
            return if let Ok(error_resp) = serde_json::from_str::<FIBErrorResponse>(&response_text) {
                Err(MesopotamiaError::from_gateway_code(
                    "fib",
                    &error_resp.error,
                    &error_resp.error_description,
                ))
            } else {
                Err(MesopotamiaError::GatewayError {
                    provider: "fib".to_string(),
                    code: status.as_u16().to_string(),
                    message: response_text,
                })
            };
        }

        let token_response: FIBTokenResponse = serde_json::from_str(&response_text)
            .map_err(|e| MesopotamiaError::SerializationError(e))?;

        let expires_at = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .map_err(|e| MesopotamiaError::Unknown(e.to_string()))?
            .as_secs() + token_response.expires_in - 300; // Refresh 5 minutes before expiry

        Ok(FIBTokenInfo {
            access_token: token_response.access_token,
            token_type: token_response.token_type,
            expires_at,
        })
    }

    /// Verify webhook signature
    pub fn verify_webhook(&self, signature: &str, payload: &str) -> bool {
        generate_hmac_sha256(&self.client_secret, payload)
            .map(|sig| format!("sha256={}", sig) == signature)
            .unwrap_or(false)
    }

    /// Handle error response from gateway
    fn handle_error(&self, status: reqwest::StatusCode, body: &str) -> Result<PaymentResponse> {
        if let Ok(error_resp) = serde_json::from_str::<FIBErrorResponse>(body) {
            Err(MesopotamiaError::from_gateway_code(
                "fib",
                &error_resp.error,
                &error_resp.error_description,
            ))
        } else {
            Err(MesopotamiaError::GatewayError {
                provider: "fib".to_string(),
                code: status.as_u16().to_string(),
                message: body.to_string(),
            })
        }
    }
}

/// FIB request/response types
#[derive(Debug, Serialize)]
struct FIBCreatePaymentRequest {
    amount: u64,
    order_id: String,
    callback_url: String,
    webhook_url: String,
    #[serde(skip_serializing_if = "String::is_empty")]
    description: String,
    metadata: HashMap<String, String>,
}

#[derive(Debug, Deserialize)]
struct FIBCreatePaymentResponse {
    payment_id: String,
    status: String,
    deep_link: String,
    web_url: String,
    qr_code: String,
    valid_until: String,
}

#[derive(Debug, Deserialize)]
struct FIBPaymentStatusResponse {
    payment_id: String,
    status: String,
    amount: u64,
    order_id: String,
}

#[derive(Debug, Deserialize)]
struct FIBTokenResponse {
    access_token: String,
    token_type: String,
    expires_in: u64,
}

#[derive(Debug, Deserialize)]
struct FIBErrorResponse {
    error: String,
    error_description: String,
}

/// Cached token information
#[derive(Debug, Clone)]
struct FIBTokenInfo {
    access_token: String,
    token_type: String,
    expires_at: u64,
}

impl FIBTokenInfo {
    /// Check if token is still valid
    fn is_valid(&self) -> bool {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap_or_default()
            .as_secs();

        now < self.expires_at
    }
}

/// Parse payment status from string
fn parse_status(status: &str) -> PaymentStatus {
    match status.to_lowercase().as_str() {
        "pending" => PaymentStatus::Pending,
        "completed" => PaymentStatus::Completed,
        "failed" => PaymentStatus::Failed,
        "cancelled" | "declined" => PaymentStatus::Cancelled,
        "expired" => PaymentStatus::Expired,
        _ => PaymentStatus::Pending,
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parse_status() {
        assert_eq!(parse_status("pending"), PaymentStatus::Pending);
        assert_eq!(parse_status("completed"), PaymentStatus::Completed);
        assert_eq!(parse_status("FAILED"), PaymentStatus::Failed);
        assert_eq!(parse_status("declined"), PaymentStatus::Cancelled);
    }

    #[test]
    fn test_client_creation() {
        let client = FIBClient::new(
            "https://test.com".to_string(),
            "client_123".to_string(),
            "secret".to_string(),
            30000,
        );
        assert_eq!(client.client_id, "client_123");
        assert_eq!(client.base_url, "https://test.com");
    }

    #[test]
    fn test_token_validity() {
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();

        // Valid token (expires in 1 hour)
        let valid_token = FIBTokenInfo {
            access_token: "test_token".to_string(),
            token_type: "Bearer".to_string(),
            expires_at: now + 3600,
        };
        assert!(valid_token.is_valid());

        // Expired token
        let expired_token = FIBTokenInfo {
            access_token: "test_token".to_string(),
            token_type: "Bearer".to_string(),
            expires_at: now - 100,
        };
        assert!(!expired_token.is_valid());
    }
}
